### PR TITLE
LTI account linking page

### PR DESCRIPTION
**Pull request by carl-codeorg**
Screenshot 2024-05-21 at 1 48 47 PM

Adds some more of the LTI account linking functionality:
- Links a new LTI auth option to an existing account when a new LTI user successfully auths via Google during the registration flow.
- Adds a new version of the login page for the account linking flow. This page conditionally renders instead of the normal login page when the DCDO flag for account linking is enabled, and there is an in-progress partial registration for an LTI user.

Out of scope for this PR:
- Support for Microsoft and Facebook auth. If the users clicks those buttons, they will log the user in as usual but no account linking will happen.
- Support for users created via roster sync. Right now, they skip the account linking flow.

## Links

- [JIRA](https://codedotorg.atlassian.net/browse/P20-917)
- [JIRA](https://codedotorg.atlassian.net/browse/P20-910)
- [Figma](https://www.figma.com/design/88sbry7xnl1wcyrOgbgFtn/Sign-Up-Flow?node-id=2667-8615&t=Xs9sYUrsM92VQJzf-0)

## Testing story

To test this locally, some extra configuration is required. Google oauth requires https to work with non-localhost domains, such as `localhost-studio.code.org`. Since Canvas launches need to work as well, Canvas also needs to be configured to launch into a local Code.org server using https. Steps are as follows:
1. Configure locals.yml to run https:
```
https_development: true
default_scheme: 'https:'
dashboard_port: 443
```
2. Run dashboard-server while passing the --ssl flag to `thin`. This can be accomplished by adding `--ssl` to the arguments to thin start on [line 19 of the dashboard-server script](https://github.com/code-dot-org/code-dot-org/blob/staging/bin/dashboard-server#L19). There's also a [PR up](https://github.com/code-dot-org/code-dot-org/pull/58821) to make this happen automatically.
3. Configure your Canvas developer key with the proper https URLs:
```
Redirect URI: https://localhost-studio.code.org/lti/v1/authenticate
Target Link URI: https://localhost-studio.code.org/lti/v1/sync_course
OIDC Initiation URI: https://localhost-studio.code.org/lti/v1/login
```
4. GCP has already been configured to work with the localhost https URL. As long as your local server is listening on port 443, it should work.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx [hunk 1]
```diff
@@ -14,7 +14,12 @@ import i18n from '@cdo/locale';
 import {LtiProviderContext} from '../../context';
 
 const ExistingAccountCard = () => {
-  const {ltiProviderName, existingAccountUrl} = useContext(LtiProviderContext)!;
+  const {ltiProvider, ltiProviderName, existingAccountUrl} =
+    useContext(LtiProviderContext)!;
+  const urlParams = new URLSearchParams({
+    lms_name: ltiProviderName,
+    lti_provider: ltiProvider,
+  });
```
Comment (stephenliang) [inline apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx]: What do you think of this?

```
diff --git a/apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx b/apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx
index d8d8a6b2fea..f0f70b9ca52 100644
--- a/apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx
+++ b/apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx
@@ -16,10 +16,12 @@ import {LtiProviderContext} from '../../context';
 const ExistingAccountCard = () => {
   const {ltiProvider, ltiProviderName, existingAccountUrl} =
     useContext(LtiProviderContext)!;
+  const callToActionUrl = new URL(existingAccountUrl);
   const urlParams = new URLSearchParams({
     lms_name: ltiProviderName,
     lti_provider: ltiProvider,
   });
+  callToActionUrl.search = urlParams.toString();
 
   return (
     <Card data-testid={'existing-account-card'}>
@@ -43,7 +45,7 @@ const ExistingAccountCard = () => {
           color={buttonColors.purple}
           type={'primary'}
           size="l"
-          href={`${existingAccountUrl}?${urlParams.toString()}`}
+          href={callToActionUrl.href}
           text={i18n.ltiLinkAccountExistingAccountCardActionLabel()}
         />
       </CardActions>
diff --git a/apps/test/unit/lib/ui/lti/link/LtiLinkAccountPage/LtiLinkAccountPageTest.tsx b/apps/test/unit/lib/ui/lti/link/LtiLinkAccountPage/LtiLinkAccountPageTest.tsx
index 10237f182e7..5abcd27da68 100644
--- a/apps/test/unit/lib/ui/lti/link/LtiLinkAccountPage/LtiLinkAccountPageTest.tsx
+++ b/apps/test/unit/lib/ui/lti/link/LtiLinkAccountPage/LtiLinkAccountPageTest.tsx
@@ -12,7 +12,7 @@ const DEFAULT_CONTEXT: LtiProviderContextProps = {
   ltiProvider: 'canvas_cloud',
   ltiProviderName: 'Canvas',
   newAccountUrl: '/new-account',
-  existingAccountUrl: '/existing-account',
+  existingAccountUrl: 'https://test.code.org/existing-account',
 };
 
 describe('LTI Link Account Page Tests', () => {
@@ -41,7 +41,9 @@ describe('LTI Link Account Page Tests', () => {
           .getByText(i18n.ltiLinkAccountExistingAccountCardActionLabel())
           .closest('a')!
           .getAttribute('href')
-      ).to.equal('/existing-account?lms_name=Canvas&lti_provider=canvas_cloud');
+      ).to.equal(
+        'https://test.code.org/existing-account?lms_name=Canvas&lti_provider=canvas_cloud'
+      );
     });
   });
```

Comment (carl-codeorg) [reply apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx]: Nice, much cleaner than cludging the url together in a template string!

**Review by stephenliang — APPROVED** [review]

File: apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx (hunk 1 above)
Comment (carl-codeorg) [reply apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx]: @stephenliang What do you think about [this](https://github.com/code-dot-org/code-dot-org/pull/58823/commits/acc1a870c5bb2bdb7d2d04de7909d31f0428e6ec)? Makes existingAccountUrl a URL to begin with in the context props. Cleaner?

Comment (stephenliang) [reply apps/src/lib/ui/lti/link/LtiLinkAccountPage/cards/ExistingAccountCard/index.tsx]: That looks even better!

**Review by nicklathe — APPROVED** [review]

**Pull request merged**